src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs

```diff
@@ -431,28 +432,142 @@ private static bool OverloadResolutionResultIsValid<TMember>(ArrayBuilder<Member
                 if (result.Result.IsValid)
                 {
-                    if (!member.CallingConvention.IsCallingConvention(expectedConvention)
-                        || member.CallingConvention.HasUnknownCallingConventionAttributeBits())
+                    if (member.CallingConvention.HasUnknownCallingConventionAttributeBits())
                     {
-                        results[i] = new MemberResolutionResult<TMember>(
-                            result.Member, result.LeastOverriddenMember,
-                            MemberAnalysisResult.WrongCallingConvention());
+                        results[i] = makeWrongCallingConvention(result);
+                        continue;
+                    }
+
+                    // We're not in an attribute, so cycles shouldn't be possible
+                    var unmanagedCallersOnlyData = member.GetUnmanagedCallersOnlyAttributeData(forceComplete: true);
+
+                    Debug.Assert(!ReferenceEquals(unmanagedCallersOnlyData, UnmanagedCallersOnlyAttributeData.AttributePresentDataNotBound)
+                                 && !ReferenceEquals(unmanagedCallersOnlyData, UnmanagedCallersOnlyAttributeData.Uninitialized));
+
+                    Cci.CallingConvention actualCallKind;
+                    ImmutableHashSet<INamedTypeSymbolInternal> actualUnmanagedCallingConvetionTypes;
+
+                    if (unmanagedCallersOnlyData is null)
+                    {
+                        actualCallKind = member.CallingConvention;
+                        actualUnmanagedCallingConvetionTypes = ImmutableHashSet<INamedTypeSymbolInternal>.Empty;
+                    }
+                    else
+                    {
+                        // There's data from an UnmanagedCallersOnlyAttribute present, with takes precedence over the
+                        // CallKind bit in the method definition. We use the following rules to decode the attribute:
+                        // * If no types are specified, the CallKind is treated as Unmanaged, with no unmanaged calling convention types
+                        // * If there is one type specified, and that type is named CallConvCdecl, CallConvThiscall, CallConvStdcall, or 
+                        //   CallConvFastcall, the CallKind is treated as CDecl, ThisCall, Standard, or FastCall, respectively, with no
+                        //   calling types.
+                        // * If multiple types are specified or the single type is not named one of the specially called out types above,
+                        //   the CallKind is treated as Unmanaged, with the union of the types specified treated as calling convention types.
+
+                        var unmanagedCallingConventionTypes = unmanagedCallersOnlyData.CallingConventionTypes;
+                        Debug.Assert(unmanagedCallingConventionTypes.All(u => FunctionPointerTypeSymbol.IsCallingConventionModifier((NamedTypeSymbol)u)));
```

Comment:
>u [](start = 149, length = 1)

Could we have a null in the set? It looks like ```IsCallingConventionModifier``` isn't ready to deal with nulls.

Reply:
No. In source, a `null` won't get added. And from metadata, seeing a `null` will cause us to treat the data as invalid and issue a use-site diagnostic on the use of `M`.

---
In reply to: [491113023](https://github.com/dotnet/roslyn/pull/47651#discussion_r491113023) [](ancestors = 491113023)

Reply:
Added an additional test showing metadata behavior around null.

---
In reply to: [491121238](https://github.com/dotnet/roslyn/pull/47651#discussion_r491121238) [](ancestors = 491121238,491113023)